### ISSUE TITLE
add animation for reset action

### ISSUE THREAD

**Issue by sluramod** [react-community/react-navigation#351]
When using reset action there should be animation when transitioning to the new screen.

**Comment by ph3b** [comment]
+1

**Comment by grabbou** [comment]
Can you provide more info what you mean? What action are you dispatching that there is no animation? AFAIK there's animation when transitioning from old state to new state (and some of the users requested it to be removed and implement transition-less reset).

**Comment by ph3b** [comment]
I think OP means that when you dispatch an action like so:
```javascript
this.props.navigation.dispatch({
        type: 'Navigation/RESET',
        index: 0,
        actions: [{ type: 'Navigate', routeName: 'Home' }]
})
```
it should still be possible to use the animation transition (slide from left or right). This is something that I would love to see as well.

**Comment by grabbou** [comment]
The animations are designed to only happen at the index change and so, replacing screen shouldn't produce any animations because functionally, there's nothing to animate. You just replaced scene at a given index. Maybe once we have added custom transitions, it's possible.

For now, I'd suggest, assuming you want to replace animated `A` with `B` and your stack is `A`, reset it to `[B, A]` and call `pop`.

**Comment by ph3b** [comment]
Ah, I see.
I'm trying something like this, but it doesn't quite work. Is this what you mean?

```javascript
 this.props.navigation.dispatch({
        type: 'Navigation/RESET',
        index: 0,
        actions: [
          { type: 'Navigate', routeName: 'Waiting', params: { payload } },
          { type: 'Navigate', routeName: 'Home' }
        ]
      })
this.props.navigation.goBack();
```

**Comment by brunitob** [comment]
any workaround?

**Comment by skyworld** [comment]
I have the same problem  tooo!

**Comment by ashemah** [comment]
Any news on this functionality? The reset command should have an animation in the same way as navigating to a new Activity and resetting the nav stack on Android has an animation and setting the ViewController stack array manually on iOS also has an animation. 

React Navigation's reset functionality currently doesn't act like any of the real mobile platform navigation stacks.

**Comment by apperside** [comment]
+1

**Comment by eliprodigy** [comment]
+1

**Comment by edencakir** [comment]
+1

**Comment by meilers** [comment]
+1

**Comment by davidysong** [comment]
+1

**Comment by Fendrian** [comment]
For my use case (a loading splash screen with a nice transition that users can't go back to), one workaround that seems effective is to ignore the idea resetting entirely and instead just conceptually assign some screens as 'top-level', and change the back button functionality so that the app exits when back is pressed on those screens.

```
Nav = StackNavigator(navScreens, navOptions);
const defaultGetStateForAction = Nav.router.getStateForAction;
Nav.router.getStateForAction = (action, state) => {
  if (
    state &&
    action.type === NavigationActions.BACK &&
    (
      state.routes[state.index].routeName === 'Login' ||
      state.routes[state.index].routeName === 'Main'
    )
  ) {
    // Returning null indicates stack end, and triggers exit
    return null;
  }
  return defaultGetStateForAction(action, state);
};

```
As is mentioned in #1372 (the inspiration for this workaround), returning 'null' from the back action is the signal that the end of the stack has been reached and it's time to exit. So with the code above, whenever the back button is pressed on the 'Login' or 'Main' screens, the user experience is the same as it would be at index 0 - the app exits, which also clears the navigation stack so it's fresh when they re-launch.

I know this doesn't directly address the issue, but it might help someone who is here looking for a quick workaround in the meantime.

**Comment by faceyspacey** [comment]
I think the issue is more nuanced than just animating resets. **For one, it already does, as long as you have a different number of routes in the new stack.** However, this isn't correct either because if there is a big difference in the number of routes, you see a lot of pages animated sliding by, when really the user likely only expects to perceive one page slide by. 

In addition, the OP wants to handle a particular scenario where the same number of routes are reset on the stack, in which case there is no animation (it's just "replaced"). The solution of dispatching a reset and then following that up with a second dispatch (perhaps pushing a new route or going back) actually may be fine as a long term solution.

The optimum solution for all however is essentially a key in the action to designate whether it should be animated or not.  And it automatically determines whether it's a "back" animation or a "forward" animation based on whether there are now less or more routes on the stack (which it already does). Though perhaps that also should be customizable, as for example, even though the new stack may have less routes, the desired user experience may still be a forward animation; and now when the user goes back--likely unbeknownst to him--there are less routes to move back through.

Either way, the biggest problem here in my opinion is that the transitioner slides through more than one route if the difference in routes is more than 1. It doesn't animate perfectly; the text in the nav bar gets stuck, etc. But also just seeing that many pages slide by doesn't feel right. Since it already automatically animates forward or back based on the difference in routes, it's my opinion the first fix should be simply animating a single always.

**Comment by faceyspacey** [comment]
There's also another key issue: when you reset a stack in a family of navigators/routers, the UI automatically takes you to that stack (perhaps changing tabs, closing drawers, etc). That's a problem too. Just imagine you have a TabNavigator with StackNavigators on each tab. The user may have done an action such as completing a purchase or creating some content, and you want to reset some of the other tabs, but without taking the user out of the current tab. You just want it so that *when he returns himself* to those sections of the app they are reset. You don't want to force him to go there.

**Comment by faceyspacey** [comment]
I posted the final (hopefully) solution I came up with in a new issue that synthesizes everything here clearly:

https://github.com/react-community/react-navigation/issues/1663

**Comment by Tom29** [comment]
This is weird, i still can see the animation after resetting (move forward)
I want to remove it. because i want it go to the level i wanted at beginning #2086 
No matter what i've tried, i still see the animation, it's ugly and annoying ...

**Comment by kelset** [comment]
Closing this in favor of #1663 (since also the issue has been quiet for a long time too).

**Comment by blaues0cke** [comment]
@Fendrian Nice explanation! Just tried your workaround but it seems the "swipe back" gesture on iOS does still work:

![nov-05-2017 11-34-07](https://user-images.githubusercontent.com/5159398/32413998-42b14d64-c21d-11e7-8541-27a7e679d9de.gif)

**Comment by filbertteo** [comment]
@blaues0cke use `gesturesEnabled: false` for that screen's `navigationOptions`, e.g.

```
IntroScreen: {
  screen: IntroScreen,
  navigationOptions: {
    gesturesEnabled: false
  }
}
```

**Comment by mbret** [comment]
I wanted to give you my feedback after having spent my evening trying to have animation while resetting.

First here is some context:

I have an app which use redux, rxjs and redux storage. I have a splashcreen as first route (not the splashcreen that you can use before app load, an actual screen) which display a loader while redux storage load asynchronously my store. Once the store is ready I redirect from the splash to my Home. Since now I was using Navigation.reset to go to the Home as index 0 and get away of my splash screen. The Splashscreen has no more purpose and there should be no way to go back on it anyway, that's why I use reset and remove this screen.

Now the problem is that from Splash to Home this is kind of "brutal" without animation and I wanted a "fade" like animation. I tried so to use Navigation.navigate instead and it works as expected with animation. By using specific logic to backPress and overwriting the getStateForAction I was able to completely block the navigation back to Splash and exit the app when using backPress on Home as expected. The index 0 from the route stack is a kind of "unaccessible / ignored route"

The Splash screen is now always present in the routes stack and it also means that when exiting the app with backPress and resuming the app, all the screen in route stack get mounted back (usually, Splash and Home). My SplashScreen is mount a new time and the logic inside as well. Of course I could handle it and check if the app has already been started before and etc but this start to give me a lot of boilerplate and weird code to handle this problem. This is a workaround which pollute my app for no reason and does not seems "right"

So yes having an animation with reset is a need for me and sounds completely fine.

**Comment by mateusz-as** [comment]
My solution:
- after initialization i use reset for "hard" replace splash screen to home
- my initial home view have 'copy' of splash screen layout on top of other ui
- fade out splash screen ui copy

Its also dirty , but at the end you have little less pollution (only one additional invisible view).

**Comment by jitenderchand1** [comment]
is anyone has fixed this?

**Comment by hb2708** [comment]
I settled with the following approach as react-navigation as of now still missing this.

Just add a animation to the root component(or component you want to slide in)
```
state = {
        parentViewX: new Animated.Value(width),
    };

componentDidMount() {
        this.slideInParentView();
}

slideInParentView() {
    Animated.spring(this.state.parentViewX, {
            toValue: 0,
        }).start();
    }

render() {
            return (
                <Animated.View
                    style={ [{your_styles}, { transform: [{ translateX: this.state.parentViewX }] }] }>
   </Animated.View>
            )
        }
                    
```

**Comment by farshidshahmoradi1996** [comment]
I have same problem in using react-navigation 
this.props.navigation.replace or this.props.navigation.reset has no effect :(

**Comment by farshidshahmoradi1996** [comment]
> @jitenderchand1
> I settled with the following approach as react-navigation as of now still missing this.
> 
> Just add a animation to the root component(or component you want to slide in)
> 
> ```
> state = { parentViewX: new Animated.Value(width) };
> 
> componentDidMount() { this.slideInParentView(); }
> 
> slideInParentView() {
>     Animated.spring(this.state.parentViewX, { toValue: 0 }).start();
> }
> 
> render() {
>     return (
>         <Animated.View style={ [{your_styles}, { transform: [{ translateX: this.state.parentViewX }] }] }>
>         //Your views/components
>         </Animated.View>
>     )
> }
>                     
> ```

whats width here ??
